server/src/main/java/org/folio/shared/index/api/OaiService.java

```diff
@@ -0,0 +1,374 @@
+        errorCode = "internal";
+      }
+      ctx.response().write("  <error code=\"" + errorCode + "\">"
+          + XmlJsonUtil.encodeXmlText(e.getMessage()) + "</error>\n");
+      oaiFooter(ctx);
+      return Future.succeededFuture();
+    });
+  }
+
+  static Future<Void> getCheck(RoutingContext ctx) {
+    try {
+      RequestParameters params = ctx.get(ValidationHandler.REQUEST_CONTEXT_KEY);
+      String verb = Util.getParameterString(params.queryParameter("verb"));
+      if (verb == null) {
+        throw OaiException.badVerb("missing verb");
+      }
+      String metadataPrefix = Util.getParameterString(params.queryParameter("metadataPrefix"));
+      if (metadataPrefix != null && !"marcxml".equals(metadataPrefix)) {
+        throw OaiException.cannotDisseminateFormat("only metadataPrefix \"marcxml\" supported");
+      }
+      switch (verb) {
+        case "ListIdentifiers":
+          return listRecords(ctx, false);
+        case "ListRecords":
+          return listRecords(ctx, true);
+        case "GetRecord":
+          return getRecord(ctx);
+        default:
+          throw OaiException.badVerb(verb);
+      }
+    } catch (Exception e) {
+      return Future.failedFuture(e);
+    }
+  }
+
+  static Future<Void> listRecords(RoutingContext ctx, boolean withMetadata) {
+    RequestParameters params = ctx.get(ValidationHandler.REQUEST_CONTEXT_KEY);
+    String set = Util.getParameterString(params.queryParameter("set"));
+    if (set == null) {
+      throw OaiException.badArgument("set or resumptionToken missing");
+    }
+    Storage storage = new Storage(ctx);
+    return storage.selectMatchKeyConfig(set).compose(conf -> {
+      if (conf == null) {
+        throw OaiException.badArgument("set \"" + set + "\" not found");
+      }
+      List<Object> tupleList = new ArrayList<>();
+      tupleList.add(set);
+      StringBuilder sqlQuery = new StringBuilder("SELECT * FROM " + storage.getClusterMetaTable()
+          + " WHERE match_key_config_id = $1");
+      String from = Util.getParameterString(params.queryParameter("from"));
+      int no = 2;
+      if (from != null) {
+        tupleList.add(LocalDateTime.parse(from));
+        sqlQuery.append(" AND modified >= $" + no);
+        no++;
+      }
+      String until = Util.getParameterString(params.queryParameter("until"));
+      if (until != null) {
+        tupleList.add(LocalDateTime.parse(until));
+        sqlQuery.append(" AND modified <= $" + no);
+      }
+      return storage.getPool().getConnection().compose(conn ->
+          listRecordsResponse(ctx, storage, conn, sqlQuery.toString(), Tuple.from(tupleList),
+              withMetadata)
+      );
+    });
+  }
+
+  private static void endListResponse(RoutingContext ctx, SqlConnection conn, Transaction tx,
+      String elem) {
+    ctx.response().write("  </" + elem + ">\n");
+    oaiFooter(ctx);
+    tx.commit().compose(y -> conn.close());
+  }
+
+  static void clearMarcField(JsonObject marc, String tag) {
```

Comment:
can we factor out Marc-constructing methods to its own class?

Reply:
That's actually now done.